pkg/controller/core.oam.dev/v1alpha2/applicationrollout/rollout_handler.go

```diff
@@ -0,0 +1,236 @@
+		h.appRollout.Status.RollingState != v1alpha1.RolloutFailedState {
+		// happen when roll forward or revert in middle of rollout, previous rollout haven't finished
+		// continue to handle the previous resources until we are okay to move forward
+		h.targetRevName = h.appRollout.Status.LastUpgradedTargetAppRevision
+		h.sourceRevName = h.appRollout.Status.LastSourceAppRevision
+	} else {
+		// previous rollout have finished, go ahead using new source/target revision
+		h.targetRevName = h.appRollout.Spec.TargetAppRevisionName
+		h.sourceRevName = h.appRollout.Spec.SourceAppRevisionName
+		// mark so that we don't think we are modified again
+		h.appRollout.Status.LastUpgradedTargetAppRevision = h.appRollout.Spec.TargetAppRevisionName
+		h.appRollout.Status.LastSourceAppRevision = h.appRollout.Spec.SourceAppRevisionName
+	}
+	h.appRollout.Status.StateTransition(v1alpha1.RollingModifiedEvent)
+}
+
+// templateTargetManifest call dispatch to template target manifest to cluster
+func (h *rolloutHandler) templateTargetManifest(ctx context.Context) error {
+	var rt *v1beta1.ResourceTracker
+	// only when sourceAppRevision is not nil, we need gc old revision resources
+	if h.sourceAppRevision != nil {
+		rt := new(v1beta1.ResourceTracker)
+		err := h.Get(ctx, types.NamespacedName{Name: dispatch.ConstructResourceTrackerName(h.appRollout.Spec.SourceAppRevisionName, h.appRollout.Namespace)}, rt)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+
+	// use source resourceTracker to handle same resource owner transfer
+	dispatcher := dispatch.NewAppManifestsDispatcher(h, h.targetAppRevision).EnableUpgradeAndSkipGC(rt)
+	_, err := dispatcher.Dispatch(ctx, h.targetManifests)
+	if err != nil {
+		return err
+	}
+	workload, err := h.extractWorkload(ctx, *h.targetWorkloads[h.needRollComponent])
+	if err != nil {
+		return err
+	}
+	ref := metav1.GetControllerOfNoCopy(workload)
+	if ref.Kind == v1beta1.ResourceTrackerKind {
+		wlPatch := client.MergeFrom(workload.DeepCopy())
+		// guarantee resourceTracker isn't controller owner of workload
+		disableControllerOwner(workload)
+		if err = h.Client.Patch(ctx, workload, wlPatch, client.FieldOwner(h.appRollout.UID)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// handle rollout succeed work left
+func (h *rolloutHandler) finalizeRollingSucceeded(ctx context.Context) error {
+	var rt *v1beta1.ResourceTracker
+	// only when sourceAppRevision is not nil, we need gc old revision resources
+	if h.sourceAppRevision != nil {
+		rt := new(v1beta1.ResourceTracker)
+		err := h.Get(ctx, types.NamespacedName{Name: dispatch.ConstructResourceTrackerName(h.appRollout.Spec.SourceAppRevisionName, h.appRollout.Namespace)}, rt)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+	}
+	// dispatch here gc useless workloads/traits
+	dispatcher := dispatch.NewAppManifestsDispatcher(h, h.targetAppRevision).EnableGC(rt)
```

Comment:
The `finalizeSucceed` logic remains to be revisited. 

-  We must change the workload's controller-owner back to `ResourceTracker` here.
- After finalizing succeed, the source revision's resource tracker is supposed to deleted. But if there's any rollout event is requeued again, it will run into finalizing succeed logic according to rollout status. Since source revision's resource tracker is already deleted, this round of finalization will fail for dispatch cannot get previous resource tracker. An error will be returned, which will trigger next round of reconciliation. Endless looping.